File: src/cpp_data_structures/data_structures.cpp
#include "data_structures.h"

CtQueue::CtQueue(int qz): q_size(qz)
{
  front = rear = nullptr;
  items = 0;
}

CtQueue::~CtQueue()
{
  Node * temp;
  while (front != nullptr)
  {
    temp = front;
    front = front->next;
    delete temp;
  }
}

bool CtQueue::is_empty() const
{
  return items == 0;
}

bool CtQueue::is_full() const
{
  return items == q_size;
}

int CtQueue::count() const
{
  return items;
}

bool CtQueue::enqueue(const Item & item)
{
  if (is_full())
  {
    return false;
  }

  Node * add = new Node;
  add->item = item;
  add->next = nullptr;

  if (front == nullptr)
  {
    front = add;
  }
  else
  {
    rear->next = add;
  }

  items++;

  rear = add;
  return true;
}

bool CtQueue::dequeue(Item & item)
{
  if (is_empty())
  {
    return false;
  }

  Node * temp = front;
  front = front->next;
  item = temp->item;
  delete temp;
  items--;

  if (front == nullptr)
  {
    rear = nullptr;
  }

  return true;
}

File: chocotrade/core/data_structures.py
"""
抽象数据结构
"""
from queue import Empty as Empty
from queue import Queue as Queue


class CtQueue:
    """"""
    MAX_SIZE = 10

    def __init__(self, _max_size: int = MAX_SIZE):
        """"""
        self._max_size = _max_size
        self._count = 0
        self.front = None
        self.rear = None

    class Node:
        """"""
        def __init__(self, item: int):
            """"""
            self.item = item
            self.next = None

    def count(self):
        """"""
        return self._count

    def is_empty(self) -> bool:
        """"""
        return self._count == 0

    def is_full(self) -> bool:
        """"""
        return self._count == self.max_size

    def enqueue(self, x: int) -> bool:
        """"""
        if self.is_full():
            return False

        add = self.Node(x)
        if self.front is None:
            self.front = add
        else:
            self.rear.next = add

        self.rear = add
        return True

    def deqeueue(self) -> tuple[bool, int]:
        """"""
        if self.is_empty():
            return (False, None)

        temp = self.front.item
        self.front = self.front.next

        if self.front is None:
            self.rear = None

        return (True, temp)
